src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/InvocationMessage.cs

```diff
@@ -36,6 +37,29 @@ public InvocationMessage(string invocationId, string target, object[] arguments,
             NonBlocking = nonBlocking;
         }
 
+        public override bool Equals(object obj)
+        {
+            return obj is InvocationMessage m && Equals(m);
+        }
+
+        public override int GetHashCode()
+        {
+            var combiner = new HashCodeCombiner();
+            combiner.Add(InvocationId);
+            combiner.Add(Target);
+            combiner.Add(Arguments);
+            combiner.Add(NonBlocking);
+            return combiner.CombinedHash;
+        }
+
+        public bool Equals(InvocationMessage other)
+        {
+            return string.Equals(InvocationId, other.InvocationId, StringComparison.Ordinal) &&
```

Comment:
You may want to do reference comparison first - should also take care for the null case.